Add RFC for HTLC (#220)
tari-project/tari: RFC/src/RFC-0230_HTLC.md

without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

## Goals

This document describes how to implement hashed time locked contracts as an extension to [mimblewimble](MimbleWimble).

## Related RFCs
## Description
### Abstract
Hashed Time Locked Contracts are time locked contracts that only pays out after a certain criteria has been met or refunds the originator if a certain period has expired. 

#### Time Locked contracts
In [mimblewimble](MimbleWimble) timelocked contracts can be accomplished by modifying the kernel of each transaction to include a block height. This allows a limit to how early in the blockchain lifetime the specific transaction can be included in a block. 

This requires users constructing a transaction to:
* MUST include a lockheight in the kernal of their transaction,
* MUST include the lockheight in the signed message to stop tampering with the lockheight after the transaction was constructed.

> This add the following requirement to a miner:

Reviewer: ```suggestion
Tari Miners:
```

> * MUST not add any transaction to the mined [block](block) that has not already exceeded its lockheight.

Reviewer: ```suggestion
* MUST not add any transaction to the mined [block](block) that has not already exceeded its lock height.
```

This also adds the following requirement to a [base node](base node):

> * MUST reject any [block](block) that contains a kernel with a lockheight not already past the [currenthead](currenthead).

Reviewer: ```suggestion
* MUST reject any [block](block) that contains a kernel with a lock height greater than the [currenthead](currenthead).
```

> 

Reviewer: Presumably miners and nodes MUST validate the signature against a message containing the lock height specified in the kernel as well.

Author: yes, but thats covered in the block consensus rules

### Mimblewimble N-of-N Multisig UTXO

> A normal Mimblewimble UTXO does not have a notion of being a [multisig] UTXO. The value of UTXO is hidden inside the commitment `C(v,r) = r·G + v·H` by virtue of being blinded. However, the blinding factor `r` can be composed of multiple blinding factors where `r = r1 + r2 + ... + rn`. The output commitment can then be constructed as `C(v,r) = r1·G + r2·G + ... + rn·G + v·H = (r1 + r2 + ... + rn)·G + v·H` where each participant keeps their private blinding factor hidden and only provides their public blinding factor. A multi-party aggregated signature scheme like Musig may be employed to sign such a transaction so that all parties' interests can be protected.

Reviewer: ```suggestion
A normal Mimblewimble UTXO does not have a notion of being a [multisig] UTXO. The value of the UTXO is hidden inside the commitment `C(v,r) = r·G + v·H` by virtue of being blinded. However, the blinding factor `r` can be composed of multiple blinding factors where `r = r1 + r2 + ... + rn`. The output commitment can then be constructed as `C(v,r) = r1·G + r2·G + ... + rn·G + v·H = (r1 + r2 + ... + rn)·G + v·H` where each participant keeps their private blinding factor hidden and only provides their public blinding factor. A multi-party aggregated signature scheme like MuSig may be employed to sign such a transaction so that all parties' interests can be protected.
```

> The base layer is oblivious as to how the commitment and related signature were constructed. To open such commitments (in order to spend it) only the n-of-n blinding factor `r` is required. The parties that wants to open the commitment needs to collaborate n-of-n blinding factor `r`.

Reviewer: ```suggestion
The base layer is oblivious to how the commitment and related signature were constructed. To open such commitments (in order to spend it) only the n-of-n blinding factor `r` is required. The parties that want to open the commitment need to collaborate to reconstruct the n-of-n blinding factor `r`.
```

#### Hashed Time Locked Contract
Unlike Bitcoin where this can be accomplished with a single transaction, in [MimbleWimble](MimbleWimble) this is a multi-step process to construct a timelocked contract. 

The steps are as follows:

> * The sender MUST pay all the funds into a n-of-n [multisig](multisig) [utxo](UTXO).  

Reviewer: Before getting into this it would be useful to describe what the net result of this contract is, i.e.

Alice is paying N parties, but x happens if... and y happens if..

> * All parties invlovled MUST construct a refund [transaction](transaction) paying back all funds to the sender, paying from the n-of-n [multisig](multisig) [utxo](UTXO). This [transaction](transaction) has a [transaction lock height](#Time Locked contracts) in the future and cannot be immediately mined. It therefor lives in the [mempool](mempool).

Reviewer: ```suggestion
* All parties involved MUST construct a refund [transaction] paying back all funds to the sender, spending this n-of-n [multisig] [utxo]. However, this [transaction] has a [transaction lock height](#Time Locked contracts) set in the future and cannot be immediately mined. It therefore lives in the [mempool](mempool). This means that if anything goes wrong from here on out, the sender will get his money back after the time lock expires.
```

* The sender MUST publish both above [transactions](transaction) at the same time to ensure the receiver cannot hold him hostage. 

> * The parties MUST construct a second [transaction](transaction) that pays the receiver the funds. This [transaction](transaction) typically makes use of a preimage to allow spending of the [transaction](transaction) if the user reveals some knowledge, allowing the user to unlock the [utxo](UTXO).

Reviewer: third transaction that spend the n-of-n UTXO to the receiver(s)...

Reviewer: > typically makes use of a preimage to allow spending of the [transaction](transaction) if the user reveals some knowledge

This is pretty vague. Can you be explicit here about how this is done in MW?
Who is "the user". Spender or receiver(s)?
And if a pre-image isn't used, why we'd use a HTLC at all?

Author: The examples in the links I added makes it more clear why you use a preimage. But It does not have to be used. Say it only pays out if you redeem some key in real life? Thats not a preimage.

## Disclaimer

The purpose of this document and its content is for information purposes only and may be subject to change or update
without notice.

This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
community. The release of this document is intended solely for review and discussion by the community regarding the
technological merits of the potential system outlined herein.

## Goals

This document describes how to implement hashed time locked contracts as an extension to [mimblewimble](MimbleWimble).

## Related RFCs
## Description
### Abstract
Hashed Time Locked Contracts are time locked contracts that only pays out after a certain criteria has been met or refunds the originator if a certain period has expired. 

#### Time Locked contracts
In [mimblewimble](MimbleWimble) timelocked contracts can be accomplished by modifying the kernel of each transaction to include a block height. This allows a limit to how early in the blockchain lifetime the specific transaction can be included in a block. 

This requires users constructing a transaction to:
* MUST include a lockheight in the kernal of their transaction,

> * Must include the lockheight in the signed message to stop tampering with the lockheight after the transaction was constructed.

Reviewer 2: Change "Must" to "MUST"
